scripts/kb.py

```diff
@@ -0,0 +1,653 @@
+        """
+        Return vectors for entities.
+        qids (str): Wiki QIDs.
+        RETURNS (Iterable[Iterable[float]]): Vectors for specified entities.
+        """
+        if not isinstance(qids, Sized):
+            qids = set(qids)
+
+        # Fetch row IDs for QIDs, resolve to vectors in Annoy index.
+        return self._get_vectors(
+            [
+                row["ROWID"]
+                for row in self._db_conn.cursor()
+                .execute(
+                    "SELECT ROWID FROM entities WHERE id in (%s)"
+                    % ",".join("?" * len(qids)),
+                    tuple(qids),
+                )
+                .fetchall()
+            ]
+        )
+
+    def get_vector(self, qid: str) -> Iterable[float]:
+        """
+        Return vector for qid.
+        qid (str): Wiki QID.
+        RETURNS (Iterable[float]): Vector for specified entities.
+        """
+        return next(iter(self.get_vectors([qid])))
+
+    @staticmethod
+    def _hash_file(file_path: Path, blocksize: int = 2**20) -> str:
+        """Generates MD5 file of hash iteratively (without loading entire file into memory).
+        Source: https://stackoverflow.com/a/1131255.
+        file_path (Path): Path of file to hash.
+        blocksize (int): Size of blocks to load into memory (in bytes).
+        RETURN (str): MD5 hash of file.
+        """
+        file_hash = hashlib.md5()
+        with open(file_path, "rb") as f:
+            buf = f.read(blocksize)
+            while buf:
+                file_hash.update(buf)
+                buf = f.read(blocksize)
+        return file_hash.hexdigest()
+
+    def _fetch_candidates_by_alias(
+        self, mentions: Tuple[Span, ...]
+    ) -> Dict[str, List[Dict[str, Union[str, int, float]]]]:
+        """Fetches candidates for mentions by fuzzily matching aliases to the mentions.
+        mentions (Tuple[Span, ..]): List of mentions for which to fetch candidates.
+        RETURN List[Dict[str, Dict[str, Union[str, int, float]]]]: List of candidates per mention, sorted by distance
+            to mention. Each candidate entry includes:
+            - entity ID,
+            - maximum prior probability over all aliases per entity,
+            - sum of occurences in Wikipedia over all over all aliases per entity,
+            - min. lexical distance over all over all aliases per entity,
+            - row ID of entity (relevant for linking to other tables).
+        """
+        # Subquery to fetch alias values for single mention.
+        mention_subquery = f"""
+            SELECT alias, 0 as distance, null as score FROM aliases_for_entities WHERE alias = ?
+            UNION
+            SELECT word, distance, score FROM aliases WHERE word MATCH ? AND distance <= {self._threshold_alias}
+        """
+
+        grouped_rows: Dict[str, List[Dict[str, Union[str, int, float]]]] = {}
+        for row in [
+            dict(row)
+            for row in self._db_conn.execute(
```

Comment:
The duplicate usage of the var `row` here is confusing, can we rename the inner-most one?

Reply:
Updated the outermost one in https://github.com/explosion/wikid/pull/4/commits/387b819cc24911c6065e9c8694dda62eb300b95d because I thought that would be more coherent (`grouped_rows` -> `grouped_row`). Feel free to reopen if you disagree, then I'll change it.